### PR TITLE
6LoWPAN fragment generation cleanup

### PR DESCRIPTION
**Pull request by simonduq**
This PR reworks the 6LoWPAN fragment generation logic to minimize the number of fragments and fix fragment count estimation. It also reworks logging overall, and removes duplicate code (fragment copy and transmission).

On `develop`, with default settings, we have 104 bytes left for 6LoWPAN payload in each 802.15.4 packet.  The first fragment has a 5-byte header, subsequent fragments a 4-byte header. Because fragment offsets are in 8 bytes increments, all but the last fragment must have an IPv6 payload multiple of 8. Currently, Contiki-NG also truncates the last fragment, unnecessarily. Below is an example ping with 451 bytes payload, resulting in 6 fragments:
```
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: sending packet len 507
01:00.593	ID:1	[INFO: 6LoWPAN   ] uip_len: 507, fragments: 6, free bufs: 7
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: 1st fragment (len 72, tag 0)
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: fragment (offset 15, len 96, tag 0)
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: fragment (offset 27, len 96, tag 0)
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: fragment (offset 39, len 96, tag 0)
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: fragment (offset 51, len 96, tag 0)
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: fragment (offset 63, len 3, tag 0)
```

With this PR, we save one fragment whenever the last has enough space left to accommodate the last bytes:
```
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: sending IPv6 packet with len 507
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: header len 48 -> 24, total len 507 -> 483, MAC max payload 104, frag_needed 1
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: fragmentation needed, fragments: 5, free queuebufs: 7
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: fragment 1/5 (tag 0, payload 72)
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: fragment 2/5 (tag 0, payload 96, offset 120)
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: fragment 3/5 (tag 0, payload 96, offset 216)
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: fragment 4/5 (tag 0, payload 96, offset 312)
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: fragment 5/5 (tag 0, payload 99, offset 408)
```

This PR also addresses a problem with fragment count estimation. Before fragmenting, we calculate how many fragments we will need, to abort early in case there aren't enough free queuebufs available. But the fragment count estimate is pessimistic, i.e. it may compute a too large value, and drop a packet that could otherwise have been sent. Below is an example, with a ping payload of 448 bytes. Estimates 6 fragments but then creates only 5:
```
04:41.424	ID:1	[INFO: 6LoWPAN   ] uip_len: 504, fragments: 6, free bufs: 7
04:41.424	ID:1	[INFO: 6LoWPAN   ] Fragmentation sending packet len 504
04:41.424	ID:1	[INFO: 6LoWPAN   ] output: 1st fragment (len 72, tag 4)
04:41.424	ID:1	[INFO: 6LoWPAN   ] output: fragment (offset 15, len 96, tag 4)
04:41.424	ID:1	[INFO: 6LoWPAN   ] output: fragment (offset 27, len 96, tag 4)
04:41.424	ID:1	[INFO: 6LoWPAN   ] output: fragment (offset 39, len 96, tag 4)
04:41.424	ID:1	[INFO: 6LoWPAN   ] output: fragment (offset 51, len 96, tag 4)
```

This PR replaces the estimate with an exact calculation of the number of fragments needed:
```
00:56.565	ID:1	[INFO: 6LoWPAN   ] output: fragmentation needed, fragments: 5, free queuebufs: 7
00:56.565	ID:1	[INFO: 6LoWPAN   ] output: fragment 1/5 (tag 0, payload 72)
00:56.565	ID:1	[INFO: 6LoWPAN   ] output: fragment 2/5 (tag 0, payload 96, offset 120)
00:56.565	ID:1	[INFO: 6LoWPAN   ] output: fragment 3/5 (tag 0, payload 96, offset 216)
00:56.565	ID:1	[INFO: 6LoWPAN   ] output: fragment 4/5 (tag 0, payload 96, offset 312)
00:56.565	ID:1	[INFO: 6LoWPAN   ] output: fragment 5/5 (tag 0, payload 96, offset 408)
```

Finally, we also include a general cleanup of the logs, for more readability. Before:
```
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: sending packet len 507
01:00.593	ID:1	[INFO: 6LoWPAN   ] IPHC: compressing dest or src ipaddr - setting CID
01:00.593	ID:1	[INFO: 6LoWPAN   ] IPHC: compressing src with context - setting CID & SAC ctx: 0
01:00.593	ID:1	[INFO: 6LoWPAN   ] Compressing first header: 43
01:00.593	ID:1	[INFO: 6LoWPAN   ] Compressing next header: 43
01:00.593	ID:1	[INFO: 6LoWPAN   ] Handling next header: 43 (len:8)
01:00.593	ID:1	[INFO: 6LoWPAN   ] Keeping the next header in this ext hdr: 58
01:00.593	ID:1	[INFO: 6LoWPAN   ] IPHC: last header could is not compressed: 58
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: header of len 24
01:00.593	ID:1	[INFO: 6LoWPAN   ] uip_len: 507, fragments: 6, free bufs: 7
01:00.593	ID:1	[INFO: 6LoWPAN   ] Fragmentation sending packet len 507
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: 1st fragment (len 72, tag 0)
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: fragment (offset 15, len 96, tag 0)
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: fragment (offset 27, len 96, tag 0)
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: fragment (offset 39, len 96, tag 0)
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: fragment (offset 51, len 96, tag 0)
01:00.593	ID:1	[INFO: 6LoWPAN   ] output: fragment (offset 63, len 3, tag 0)
01:00.739	ID:1	[INFO: 6LoWPAN   ] input: FRAG1 size 507, tag 0, offset 0)
01:00.739	ID:1	[INFO: 6LoWPAN   ] input: IPHC

01:00.739	ID:1	[INFO: 6LoWPAN   ] IPHC: CID flag set - increase header with one
01:00.739	ID:1	[INFO: 6LoWPAN   ] Uncompressing 8 + 0 => fd00::202:2:2:2
01:00.739	ID:1	[INFO: 6LoWPAN   ] Uncompressing 8 + 8 => fd00::1
01:00.739	ID:1	[INFO: 6LoWPAN   ] Next header is not compressed... next: 58
01:00.739	ID:1	[INFO: 6LoWPAN   ] IPHC: Found extension-header id=0 next=58 len=6
01:00.739	ID:1	[INFO: 6LoWPAN   ] Uncompressed EXT hdr: 0 len:6 exhdrlen:0 (calc: 8)
01:00.759	ID:1	[INFO: 6LoWPAN   ] input: FRAGN size 507, tag 0, offset 15)
01:00.759	ID:1	last_fragment?: packetbuf_payload_len 96 frag_size 507
01:00.759	ID:1	[INFO: 6LoWPAN   ] Fragsize: 96
01:00.807	ID:1	[INFO: 6LoWPAN   ] input: FRAGN size 507, tag 0, offset 27)
01:00.807	ID:1	last_fragment?: packetbuf_payload_len 96 frag_size 507
01:00.807	ID:1	[INFO: 6LoWPAN   ] Fragsize: 96
01:00.828	ID:1	[INFO: 6LoWPAN   ] input: FRAGN size 507, tag 0, offset 39)
01:00.828	ID:1	last_fragment?: packetbuf_payload_len 96 frag_size 507
01:00.828	ID:1	[INFO: 6LoWPAN   ] Fragsize: 96
01:00.853	ID:1	[INFO: 6LoWPAN   ] input: FRAGN size 507, tag 0, offset 51)
01:00.853	ID:1	last_fragment?: packetbuf_payload_len 96 frag_size 507
01:00.853	ID:1	[INFO: 6LoWPAN   ] Fragsize: 96
01:00.879	ID:1	[INFO: 6LoWPAN   ] input: FRAGN size 507, tag 0, offset 63)
01:00.879	ID:1	last_fragment?: packetbuf_payload_len 3 frag_size 507
01:00.879	ID:1	[INFO: 6LoWPAN   ] Fragsize: 3
01:00.879	ID:1	[INFO: 6LoWPAN   ] input: IP packet ready (length 507)
```

After:
```
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: sending IPv6 packet with len 507
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: header len 48 -> 24, total len 507 -> 483, MAC max payload 104, frag_needed 1
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: fragmentation needed, fragments: 5, free queuebufs: 7
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: fragment 1/5 (tag 0, payload 72)
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: fragment 2/5 (tag 0, payload 96, offset 120)
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: fragment 3/5 (tag 0, payload 96, offset 216)
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: fragment 4/5 (tag 0, payload 96, offset 312)
01:34.960	ID:1	[INFO: 6LoWPAN   ] output: fragment 5/5 (tag 0, payload 99, offset 408)

01:35.269	ID:1	[INFO: 6LoWPAN   ] input: received first element of a fragmented packet (tag 0, len 507)
01:35.269	ID:1	[INFO: 6LoWPAN   ] input: fragment (tag 0, payload 72, offset 0)
01:35.317	ID:1	[INFO: 6LoWPAN   ] input: fragment (tag 0, payload 96, offset 120)
01:35.338	ID:1	[INFO: 6LoWPAN   ] input: fragment (tag 0, payload 96, offset 216)
01:35.363	ID:1	[INFO: 6LoWPAN   ] input: fragment (tag 0, payload 96, offset 312)
01:35.392	ID:1	[INFO: 6LoWPAN   ] input: fragment (tag 0, payload 99, offset 408)
01:35.392	ID:1	[INFO: 6LoWPAN   ] input: received IPv6 packet with len 507
```

### CODE REVIEW

File: os/net/ipv6/sicslowpan.c [hunk 1]
```diff
@@ -1464,6 +1462,46 @@ send_packet(linkaddr_t *dest)
   watchdog_periodic();
 }
 /*--------------------------------------------------------------------*/
+/**
+ * \brief This function is called by the 6lowpan code to copy a fragment's
+ * payload from uIP and send it down the stack..
```
Comment (yatch) [inline os/net/ipv6/sicslowpan.c]: It's a super tiny thing, but ".." at the end must be a single period.

Comment (simonduq) [reply os/net/ipv6/sicslowpan.c]: fixed :)

**Comment by simonduq** [comment]
Why 1200 and not 1280?

**Comment by joakimeriksson** [comment]
1280 including all extension headers and IP-headers yes. I do not remember what the exact ping-payload should be to get there... and I guess it depends on the number of hops (as we add a ext header per hop in the path).

**Comment by simonduq** [comment]
Now merged and ready. BTW I did try test with larger fragments, and they currently fail (both on develop and this PR). Seems to be a buffering problem in Cooja motes serial interface. Not really sure. With embedded BR (as a cooja mote), ping with payload of 1000 bytes fail. With native BR (slip-radio as cooja mote), ping with 200 bytes fail.

**Comment by simonduq** [comment]
For both it's at the interface between linux host and cooja (tunslip6-BR, or nbr-slipradio). 6lowpan creates the fragments and passes them to MAC but then on the cooja side, not everything arrives as serial input it seems.

**Pull request merged**